lib/seahorn/PredicateAbstraction.cc

```diff
@@ -0,0 +1,783 @@
+		else if (!result)
+		{
+			outs() << "unsat";
+			HornDbModel absModel;
+			absModel.initModelFromFP(new_db, fp);
+			converter.setNewToOldPredMap(newToOldPredMap);
+			converter.setAbsDB(new_db);
+			converter.convert(absModel, oldModel);
+			LOG("pabs-debug", outs() << "FINAL RESULT:\n";);
+			//Print invariants
+			printInvars(db, oldModel);
+		}
+		else outs () << "unknown";
+		outs () << "\n";
+		Stats::stop("Pabs solve");
+	}
+
+	void HornDbModel::addDef(Expr fapp, Expr def)
+	{
+		if(isOpX<TRUE>(def) || isOpX<FALSE>(def))
+		{
+			relToDefMap.insert(std::pair<Expr, Expr>(bind::fname(fapp), def));
+			return;
+		}
+
+		std::ostringstream oss;
+		oss << bind::fname(bind::fname(fapp));
+		//For AbsModel
+		if(oss.str().find("_pabs") != -1)
+		{
+			ExprVector booleanVars;
+			HornDbUtils::get_all_booleans(def, std::back_inserter(booleanVars));
+			ExprMap boolVarsToBvarsMap;
+			for(Expr boolvar : booleanVars)
+			{
+				for(int i=0; i<bind::domainSz(bind::fname(fapp)); i++)
+				{
+					Expr arg_i = fapp->arg(i+1);
+					if(arg_i == boolvar)
+					{
+						Expr encoded_solution = bind::bvar(i, mk<BOOL_TY>(arg_i->efac()));
+						LOG("pabs-debug", outs() << "ENCODED SOLUTION: " << *encoded_solution << "\n";);
+						boolVarsToBvarsMap.insert(std::pair<Expr, Expr>(boolvar, encoded_solution));
+					}
+				}
+			}
+			def = replace(def, boolVarsToBvarsMap);
+			relToDefMap.insert(std::pair<Expr, Expr>(bind::fname(fapp), def));
+		}
+		//For OrigModel
+		else
+		{
+			ExprVector vars;
+			HornDbUtils::get_all_integers(def, std::back_inserter(vars));
+			ExprMap varIdMap;
+			for(Expr var : vars)
+			{
+				int var_id = variant::variantNum(bind::fname(bind::fname(var)));
+				Expr bvar = bind::intBVar(var_id, var->efac());
+				varIdMap.insert(std::pair<Expr, Expr>(var, bvar));
+			}
+			Expr orig_def = replace(def, varIdMap);
+			LOG("pabs-debug", outs() << "ADD BVAR DEF: " << *orig_def << "\n";);
+			relToDefMap.insert(std::pair<Expr, Expr>(bind::fname(fapp), orig_def));
+		}
+	}
+
+	Expr HornDbModel::getDef(Expr fapp)
+	{
+		std::ostringstream oss;
+		oss << bind::fname(bind::fname(fapp));
+		//AbsModel
+		if(oss.str().find("_pabs") != -1)
+		{
+			Expr rel = bind::fname(fapp);
+			Expr def = relToDefMap.find(rel)->second;
+			ExprMap bvarToBoolVarMap;
+			ExprVector bvars;
+			HornDbUtils::get_all_bvars(def, std::back_inserter(bvars));
+			for(Expr bvar : bvars)
+			{
+				int bvar_id = bind::bvarId(bvar);
+				bvarToBoolVarMap.insert(std::pair<Expr, Expr>(bvar, fapp->arg(bvar_id+1)));
+			}
+			Expr def_app = replace(def, bvarToBoolVarMap);
+			return def_app;
+		}
+		//OrigModel
+		else
+		{
+			Expr rel = bind::fname(fapp);
+			Expr def = relToDefMap.find(rel)->second;
+			Expr def_app = HornDbUtils::applyArgsToBvars(def, fapp, currentCandidates);
+			return def_app;
+		}
+	}
+
+	void HornDbModel::initModelFromFP(HornClauseDB &db, ZFixedPoint<EZ3> &fp)
+	{
+		ExprVector all_preds_in_DB;
+		for(HornRule r : db.getRules())
```

Comment:
seems like a wrong thing to rely on the rules. There can be predicates without rules. 
This also might visit the same predicate multiple times if there are multiple rules. 
Should just iterate over predicates in db and create variable names on the fly.